### ISSUE TITLE
user-defined matvec product for python version

### ISSUE THREAD

**Issue by tanlin2013** [primme/primme#29]
Dear primme dev team,

I was using release version 2.2. From there I am able to overwrite the function eigsh in file Primme.py so that it is possible to have user-defined matrix vector product, instead of providing the matrix explicitly, which can be quite huge and memory consuming.

My question is that, now I would like to catch up the master branch of primme, however I notice that python version is now wrapped using cython. Is is possible to provide user-defined matvec feature in the future release? 

best

**Comment by eromero-vlc** [comment]
Dear Tan,

It is possible to provide an user-defined matvec by creating a LinearOperator. See the doc at [scipy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.LinearOperator.html). There was no example in the documentation and the examples. I fixed in the commit bf5f052285:

```python
Adiag = np.arange(0, 100).reshape((100,1))
def Amatmat(x):
   if len(x.shape) == 1: x = x.reshape((100,1))
   return Adiag * x   # equivalent to diag(Adiag).dot(x)
A = scipy.sparse.linalg.LinearOperator((100,100), matvec=Amatmat, matmat=Amatmat)
evals, evecs = primme.eigsh(A, 3, tol=1e-6, which='LA')
```

Checkout master because I  made a couple of modifications so primme is more robust using LinearOperator.

**Comment by tanlin2013** [comment]
Dear Romero,

thanks for your answer! I will have a try on master.

just one more comment, maybe you can consider to write a small note on documentation. I guess there are also users who do not aware about the existence of scipy LinearOperator and thought as I did that the matrix A has to be given explicitly...  

best,
tan
